File: sdk/python/superdev_sdk.py
from __future__ import annotations

import asyncio
import json
import os
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Optional, Union

import httpx
import websockets


@dataclass
class SuperDevConfig:
    base_url: str = "http://localhost:8000"
    api_key: Optional[str] = None
    ws_url: Optional[str] = None
    timeout: int = 60


@dataclass
class Project:
    id: str
    name: str
    description: str = ""
    created_at: datetime
    updated_at: datetime
    owner_id: str
    settings: dict = field(default_factory=dict)


@dataclass
class Agent:
    id: str
    name: str
    type: str
    status: str
    config: dict = field(default_factory=dict)


@dataclass
class ChatMessage:
    role: str
    content: str
    metadata: dict = field(default_factory=dict)


@dataclass
class ChatResponse:
    id: str
    content: str
    model: str
    usage: dict = field(default_factory=dict)
    finish_reason: str = "stop"


@dataclass
class VerificationResult:
    task_id: str
    success: bool
    stage: str
    final_code: Optional[str] = None
    error: Optional[str] = None
    iterations: int = 0
    generation: dict = field(default_factory=dict)
    execution: dict = field(default_factory=dict)
    testing: dict = field(default_factory=dict)
    review: dict = field(default_factory=dict)
    correction: dict = field(default_factory=dict)


class SuperDevClient:
    def __init__(self, config: Optional[SuperDevConfig] = None):
        self.config = config or SuperDevConfig()
        self._client: Optional[httpx.AsyncClient] = None
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._api_key = self.config.api_key or os.getenv("SUPERDEV_API_KEY")

    async def __aenter__(self):
        self._client = httpx.AsyncClient(
            base_url=self.config.base_url,
            timeout=self.config.timeout,
            headers={"Authorization": f"Bearer {self._api_key}"} if self._api_key else {},
        )
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._client:
            await self._client.aclose()
        if self._ws:
            await self._ws.close()

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self._api_key}"} if self._api_key else {}

    async def _request(self, method: str, path: str, **kwargs) -> dict:
        if not self._client:
            self._client = httpx.AsyncClient(
                base_url=self.config.base_url,
                timeout=self.config.timeout,
                headers=self._headers(),
            )
        response = await self._client.request(method, path, **kwargs)
        response.raise_for_status()
        return response.json()

    async def health_check(self) -> dict:
        return await self._request("GET", "/api/v1/health")

    async def get_version(self) -> dict:
        return await self._request("GET", "/api/v1/version")

    async def create_project(
        self,
        name: str,
        description: str = "",
        template: str = "default",
    ) -> Project:
        data = await self._request("POST", "/api/v1/projects", json={
            "name": name,
            "description": description,
            "template": template,
        })
        return Project(**data["data"])

    async def list_projects(self) -> list[Project]:
        data = await self._request("GET", "/api/v1/projects")
        return [Project(**p) for p in data["data"]]

    async def get_project(self, project_id: str) -> Project:
        data = await self._request("GET", f"/api/v1/projects/{project_id}")
        return Project(**data["data"])

    async def delete_project(self, project_id: str) -> bool:
        await self._request("DELETE", f"/api/v1/projects/{project_id}")
        return True

    async def list_agents(self) -> list[Agent]:
        data = await self._request("GET", "/api/v1/agents")
        return [Agent(**a) for a in data["data"]]

    async def get_agent(self, agent_id: str) -> Agent:
        data = await self._request("GET", f"/api/v1/agents/{agent_id}")
        return Agent(**data["data"])

    async def execute_agent_task(
        self,
        agent_id: str,
        task: str,
        context: dict = None,
    ) -> dict:
        return await self._request("POST", f"/api/v1/agents/{agent_id}/execute", json={
            "task": task,
            "context": context or {},
        })

    async def chat(
        self,
        messages: list[ChatMessage],
        model: str = None,
        provider: str = None,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        stream: bool = False,
    ) -> Union[ChatResponse, AsyncIterator[ChatResponse]]:
        payload = {
            "messages": [{"role": m.role, "content": m.content} for m in messages],
            "model": model,
            "provider": provider,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": stream,
        }
        
        if stream:
            return self._stream_chat(payload)
        
        data = await self._request("POST", "/api/v1/chat/completions", json=payload)
        return ChatResponse(**data["data"])

    async def _stream_chat(self, payload: dict) -> AsyncIterator[ChatResponse]:
        async with self._client.stream("POST", "/api/v1/chat/stream", json=payload) as response:
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data = line[6:]
                    if data == "[DONE]":
                        break
                    import json
                    chunk = json.loads(data)
                    yield ChatResponse(
                        id=chunk.get("id", ""),
                        content=chunk.get("delta", ""),
                        model=chunk.get("model", ""),
                        finish_reason=chunk.get("finish_reason"),
                    )

    async def verify_code(
        self,
        task_description: str,
        language: str = "python",
        context: str = None,
        requirements: list[str] = None,
        existing_code: str = None,
        test_files: dict[str, str] = None,
        max_iterations: int = 3,
        provider: str = None,
    ) -> VerificationResult:
        data = await self._request("POST", "/api/v1/verify", json={
            "task_description": task_description,
            "language": language,
            "context": context,
            "requirements": requirements or [],
            "existing_code": existing_code,
            "test_files": test_files,
            "max_iterations": max_iterations,
            "provider": provider,
        })
        return VerificationResult(**data)

    async def generate_code(
        self,
        prompt: str,
        language: str = "python",
        context: str = None,
        provider: str = None,
    ) -> dict:
        return await self._request("POST", "/api/v1/verify/generate", json={
            "prompt": prompt,
            "language": language,
            "context": context,
            "provider": provider,
        })

    async def execute_code(
        self,
        code: str,
        language: str = "python",
    ) -> dict:
        return await self._request("POST", "/api/v1/verify/execute", json={
            "code": code,
            "language": language,
        })

    async def review_code(
        self,
        code: str,
        language: str = "python",
        context: str = None,
        provider: str = None,
    ) -> dict:
        return await self._request("POST", "/api/v1/verify/review", json={
            "code": code,
            "language": language,
            "context": context,
            "provider": provider,
        })

    async def create_workflow(
        self,
        name: str,
        description: str,
        steps: list[dict],
        variables: dict = None,
        tags: list[str] = None,
    ) -> dict:
        return await self._request("POST", "/api/v1/workflows", json={
            "name": name,
            "description": description,
            "steps": steps,
            "variables": variables or {},
            "tags": tags or [],
        })

    async def execute_workflow(
        self,
        workflow_id: str,
        variables: dict = None,
    ) -> dict:
        return await self._request("POST", f"/api/v1/workflows/{workflow_id}/execute", json={
            "variables": variables or {},
        })

    async def list_workflows(self, tags: list[str] = None) -> list[dict]:
        params = {}
        if tags:
            params["tags"] = ",".join(tags)
        data = await self._request("GET", "/api/v1/workflows", params=params)
        return data["data"]

    async def get_workflow(self, workflow_id: str) -> dict:
        data = await self._request("GET", f"/api/v1/workflows/{workflow_id}")
        return data["data"]

    async def delete_workflow(self, workflow_id: str) -> bool:
        await self._request("DELETE", f"/api/v1/workflows/{workflow_id}")
        return True

    async def create_knowledge_base(
        self,
        name: str,
        description: str = None,
        kb_type: str = "documentation",
        is_public: bool = False,
    ) -> dict:
        return await self._request("POST", "/api/v1/knowledge-bases", json={
            "name": name,
            "description": description,
            "type": kb_type,
            "is_public": is_public,
        })

    async def add_document(
        self,
        kb_id: str,
        title: str,
        content: str,
        source_url: str = None,
        source_type: str = None,
        language: str = None,
        tags: list[str] = None,
        metadata: dict = None,
    ) -> dict:
        return await self._request("POST", f"/api/v1/knowledge-bases/{kb_id}/documents", json={
            "title": title,
            "content": content,
            "source_url": source_url,
            "source_type": source_type,
            "language": language,
            "tags": tags or [],
            "metadata": metadata or {},
        })

    async def search_knowledge(
        self,
        query: str,
        kb_ids: list[str] = None,
        top_k: int = 10,
        similarity_threshold: float = 0.5,
    ) -> dict:
        return await self._request("POST", "/api/v1/knowledge-bases/search", json={
            "query": query,
            "knowledge_base_ids": kb_ids,
            "top_k": top_k,
            "similarity_threshold": similarity_threshold,
        })

    async def get_context(
        self,
        query: str,
        kb_ids: list[str] = None,
        max_tokens: int = 8000,
    ) -> dict:
        return await self._request("POST", "/api/v1/knowledge-bases/context", json={
            "query": query,
            "knowledge_base_ids": kb_ids,
            "max_tokens": max_tokens,
        })

    async def ingest_repository(
        self,
        kb_id: str,
        repo_path: str,
        file_patterns: list[str] = None,
        exclude_patterns: list[str] = None,
    ) -> dict:
        return await self._request("POST", f"/api/v1/knowledge-bases/{kb_id}/ingest-repo", json={
            "repo_path": repo_path,
            "file_patterns": file_patterns,
            "exclude_patterns": exclude_patterns,
        })

    async def find_similar_code(
        self,
        code_snippet: str,
        language: str = None,
        kb_ids: list[str] = None,
        top_k: int = 5,
    ) -> dict:
        return await self._request("POST", "/api/v1/knowledge-bases/similar-code", json={
            "code_snippet": code_snippet,
            "language": language,
            "knowledge_base_ids": kb_ids,
            "top_k": top_k,
        })

    async def list_plugins(
        self,
        plugin_type: str = None,
        tag: str = None,
        search: str = None,
    ) -> list[dict]:
        params = {}
        if plugin_type:
            params["plugin_type"] = plugin_type
        if tag:
            params["tag"] = tag
        if search:
            params["search"] = search
        data = await self._request("GET", "/api/v1/plugins/registry", params=params)
        return data

    async def install_plugin(self, slug: str) -> dict:
        return await self._request("POST", "/api/v1/plugins/install", json={"slug": slug})

    async def uninstall_plugin(self, slug: str) -> bool:
        await self._request("DELETE", f"/api/v1/plugins/{slug}")
        return True

    async def enable_plugin(self, slug: str) -> dict:
        return await self._request("POST", f"/api/v1/plugins/{slug}/enable")

    async def disable_plugin(self, slug: str) -> dict:
        return await self._request("POST", f"/api/v1/plugins/{slug}/disable")

    async def update_plugin_config(self, slug: str, settings: dict) -> dict:
        return await self._request("PUT", f"/api/v1/plugins/{slug}/config", json={"settings": settings})

    async def connect_websocket(self, project_id: str = None):
        ws_url = self.config.ws_url or self.config.base_url.replace("http", "ws")
        uri = f"{ws_url}/api/v1/ws"
        if project_id:
            uri += f"?project_id={project_id}"
        
        self._ws = await websockets.connect(uri)
        return self._ws

    async def subscribe_events(self, event_types: list[str]) -> AsyncIterator[dict]:
        if not self._ws:
            await self.connect_websocket()
        
        await self._ws.send(json.dumps({
            "type": "subscribe",
            "events": event_types,
        }))
        
        async for message in self._ws:
            yield json.loads(message)

    async def send_command(self, command: str, params: dict = None) -> dict:
        return await self._request("POST", "/api/v1/cli/execute", json={
            "command": command,
            "params": params or {},
        })


class SuperDevSyncClient:
    def __init__(self, config: Optional[SuperDevConfig] = None):
        self.config = config or SuperDevConfig()
        self._api_key = self.config.api_key or os.getenv("SUPERDEV_API_KEY")
        self._client: Optional[httpx.Client] = None

    def __enter__(self):
        self._client = httpx.Client(
            base_url=self.config.base_url,
            timeout=self.config.timeout,
            headers={"Authorization": f"Bearer {self._api_key}"} if self._api_key else {},
        )
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self._client:
            self._client.close()

    def _request(self, method: str, path: str, **kwargs) -> dict:
        if not self._client:
            self._client = httpx.Client(
                base_url=self.config.base_url,
                timeout=self.config.timeout,
                headers={"Authorization": f"Bearer {self._api_key}"} if self._api_key else {},
            )
        response = self._client.request(method, path, **kwargs)
        response.raise_for_status()
        return response.json()

    def health_check(self) -> dict:
        return self._request("GET", "/api/v1/health")

    def chat(self, messages: list[dict], **kwargs) -> dict:
        return self._request("POST", "/api/v1/chat/completions", json={
            "messages": messages,
            **kwargs,
        })

    def verify_code(self, task_description: str, **kwargs) -> dict:
        return self._request("POST", "/api/v1/verify", json={
            "task_description": task_description,
            **kwargs,
        })


async def create_client(config: Optional[SuperDevConfig] = None) -> SuperDevClient:
    client = SuperDevClient(config)
    await client.__aenter__()
    return client

File: sdk/typescript/superdev-sdk.ts
// SuperDev TypeScript SDK
// npm install @superdev/sdk

export interface SuperDevConfig {
  baseUrl: string;
  apiKey?: string;
  wsUrl?: string;
  timeout?: number;
}

export interface Project {
  id: string;
  name: string;
  description: string;
  createdAt: string;
  updatedAt: string;
  ownerId: string;
  settings: Record<string, any>;
}

export interface Agent {
  id: string;
  name: string;
  type: string;
  status: string;
  config: Record<string, any>;
}

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant' | 'function';
  content: string;
  name?: string;
  metadata?: Record<string, any>;
}

export interface ChatResponse {
  id: string;
  content: string;
  model: string;
  usage?: {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
    estimatedCost?: number;
  };
  finishReason: string;
}

export interface VerificationResult {
  taskId: string;
  success: boolean;
  stage: string;
  finalCode?: string;
  error?: string;
  iterations: number;
  generation?: any;
  execution?: any;
  testing?: any;
  review?: any;
  correction?: any;
}

export interface WorkflowStep {
  id?: string;
  name: string;
  stepType: string;
  config: Record<string, any>;
  dependsOn?: string[];
  maxRetries?: number;
  timeoutSeconds?: number;
  continueOnError?: boolean;
}

export interface Workflow {
  id: string;
  name: string;
  description: string;
  steps: WorkflowStep[];
  variables: Record<string, any>;
  tags: string[];
}

export interface KnowledgeBase {
  id: string;
  name: string;
  description?: string;
  type: string;
  isPublic: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface KnowledgeEntry {
  id: string;
  knowledgeBaseId: string;
  title: string;
  content: string;
  sourceUrl?: string;
  sourceType?: string;
  language?: string;
  tags: string[];
  metadata: Record<string, any>;
}

export interface SearchResult {
  entryId: string;
  chunkId: string;
  title: string;
  content: string;
  similarity: number;
  language?: string;
  tags: string[];
}

export interface Plugin {
  name: string;
  slug: string;
  version: string;
  description: string;
  author: string;
  pluginType: string;
  tags: string[];
  downloads: number;
  rating: number;
  isOfficial: boolean;
}

export interface InstalledPlugin {
  name: string;
  slug: string;
  version: string;
  status: string;
  config: Record<string, any>;
}

export class SuperDevClient {
  private config: SuperDevConfig;
  private client: any;
  private ws: WebSocket | null = null;

  constructor(config?: Partial<SuperDevConfig>) {
    this.config = {
      baseUrl: config?.baseUrl || 'http://localhost:8000',
      apiKey: config?.apiKey || process.env.SUPERDEV_API_KEY,
      wsUrl: config?.wsUrl,
      timeout: config?.timeout || 60000,
    };
  }

  private async request<T>(method: string, path: string, data?: any): Promise<T> {
    const url = `${this.config.baseUrl}${path}`;
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    
    if (this.config.apiKey) {
      headers['Authorization'] = `Bearer ${this.config.apiKey}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: data ? JSON.stringify(data) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const error = await response.json().catch(() => ({ detail: response.statusText }));
        throw new Error(`API Error: ${response.status} - ${error.detail || error.message}`);
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeoutId);
      throw error;
    }
  }

  async healthCheck(): Promise<any> {
    return this.request('GET', '/api/v1/health');
  }

  async getVersion(): Promise<any> {
    return this.request('GET', '/api/v1/version');
  }

  // Projects
  async createProject(name: string, description?: string, template?: string): Promise<Project> {
    const response = await this.request<{ success: boolean; data: Project }>(
      'POST',
      '/api/v1/projects',
      { name, description, template }
    );
    return response.data;
  }

  async listProjects(): Promise<Project[]> {
    const response = await this.request<{ success: boolean; data: Project[] }>('GET', '/api/v1/projects');
    return response.data;
  }

  async getProject(projectId: string): Promise<Project> {
    const response = await this.request<{ success: boolean; data: Project }>('GET', `/api/v1/projects/${projectId}`);
    return response.data;
  }

  async deleteProject(projectId: string): Promise<boolean> {
    await this.request('DELETE', `/api/v1/projects/${projectId}`);
    return true;
  }

  // Agents
  async listAgents(): Promise<Agent[]> {
    const response = await this.request<{ success: boolean; data: Agent[] }>('GET', '/api/v1/agents');
    return response.data;
  }

  async getAgent(agentId: string): Promise<Agent> {
    const response = await this.request<{ success: boolean; data: Agent }>('GET', `/api/v1/agents/${agentId}`);
    return response.data;
  }

  async executeAgentTask(agentId: string, task: string, context?: Record<string, any>): Promise<any> {
    return this.request('POST', `/api/v1/agents/${agentId}/execute`, { task, context });
  }

  // Chat
  async chat(
    messages: ChatMessage[],
    options?: {
      model?: string;
      provider?: string;
      temperature?: number;
      maxTokens?: number;
      stream?: boolean;
    }
  ): Promise<ChatResponse | AsyncGenerator<ChatResponse>> {
    const payload = {
      messages: messages.map(m => ({ role: m.role, content: m.content })),
      model: options?.model,
      provider: options?.provider,
      temperature: options?.temperature ?? 0.7,
      maxTokens: options?.maxTokens ?? 1024,
      stream: options?.stream ?? false,
    };

    if (options?.stream) {
      return this.streamChat(payload);
    }

    const response = await this.request<{ success: boolean; data: ChatResponse }>(
      'POST',
      '/api/v1/chat/completions',
      payload
    );
    return response.data;
  }

  async *streamChat(payload: any): AsyncGenerator<ChatResponse> {
    const response = await fetch(`${this.config.baseUrl}/api/v1/chat/stream`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(this.config.apiKey ? { 'Authorization': `Bearer ${this.config.apiKey}` } : {}),
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Stream error: ${response.status}`);
    }

    const reader = response.body?.getReader();
    const decoder = new TextDecoder();

    if (!reader) return;

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value, { stream: true });
        const lines = chunk.split('\n');

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const data = line.slice(6);
            if (data === '[DONE]') return;
            
            try {
              const parsed = JSON.parse(data);
              yield {
                id: parsed.id || '',
                content: parsed.delta || '',
                model: parsed.model || '',
                finishReason: parsed.finish_reason,
              };
            } catch (e) {
              // Ignore parse errors
            }
          }
        }
      }
    } finally {
      reader?.releaseLock();
    }
  }

  // Code Verification
  async verifyCode(params: {
    taskDescription: string;
    language?: string;
    context?: string;
    requirements?: string[];
    existingCode?: string;
    testFiles?: Record<string, string>;
    maxIterations?: number;
    provider?: string;
  }): Promise<VerificationResult> {
    const response = await this.request<{ success: boolean; data: VerificationResult }>(
      'POST',
      '/api/v1/verify',
      params
    );
    return response.data;
  }

  async generateCode(params: {
    prompt: string;
    language?: string;
    context?: string;
    provider?: string;
  }): Promise<any> {
    return this.request('POST', '/api/v1/verify/generate', params);
  }

  async executeCode(code: string, language?: string): Promise<any> {
    return this.request('POST', '/api/v1/verify/execute', { code, language });
  }

  async reviewCode(params: {
    code: string;
    language?: string;
    context?: string;
    provider?: string;
  }): Promise<any> {
    return this.request('POST', '/api/v1/verify/review', params);
  }

  // Workflows
  async createWorkflow(workflow: {
    name: string;
    description: string;
    steps: WorkflowStep[];
    variables?: Record<string, any>;
    tags?: string[];
  }): Promise<Workflow> {
    const response = await this.request<{ success: boolean; data: Workflow }>(
      'POST',
      '/api/v1/workflows',
      workflow
    );
    return response.data;
  }

  async executeWorkflow(workflowId: string, variables?: Record<string, any>): Promise<any> {
    return this.request('POST', `/api/v1/workflows/${workflowId}/execute`, { variables });
  }

  async listWorkflows(tags?: string[]): Promise<Workflow[]> {
    const params = tags ? { tags: tags.join(',') } : undefined;
    const response = await this.request<{ success: boolean; data: Workflow[] }>(
      'GET',
      '/api/v1/workflows',
      params
    );
    return response.data;
  }

  async getWorkflow(workflowId: string): Promise<Workflow> {
    const response = await this.request<{ success: boolean; data: Workflow }>(
      'GET',
      `/api/v1/workflows/${workflowId}`
    );
    return response.data;
  }

  async deleteWorkflow(workflowId: string): Promise<boolean> {
    await this.request('DELETE', `/api/v1/workflows/${workflowId}`);
    return true;
  }

  // Knowledge Base
  async createKnowledgeBase(params: {
    name: string;
    description?: string;
    type?: string;
    isPublic?: boolean;
  }): Promise<KnowledgeBase> {
    const response = await this.request<{ success: boolean; data: KnowledgeBase }>(
      'POST',
      '/api/v1/knowledge-bases',
      params
    );
    return response.data;
  }

  async addDocument(kbId: string, params: {
    title: string;
    content: string;
    sourceUrl?: string;
    sourceType?: string;
    language?: string;
    tags?: string[];
    metadata?: Record<string, any>;
  }): Promise<KnowledgeEntry> {
    const response = await this.request<{ success: boolean; data: KnowledgeEntry }>(
      'POST',
      `/api/v1/knowledge-bases/${kbId}/documents`,
      params
    );
    return response.data;
  }

  async searchKnowledge(params: {
    query: string;
    knowledgeBaseIds?: string[];
    topK?: number;
    similarityThreshold?: number;
  }): Promise<SearchResult[]> {
    const response = await this.request<{ success: boolean; data: SearchResult[] }>(
      'POST',
      '/api/v1/knowledge-bases/search',
      params
    );
    return response.data;
  }

  async getContext(params: {
    query: string;
    knowledgeBaseIds?: string[];
    maxTokens?: number;
  }): Promise<{ context: string; totalTokens: number }> {
    return this.request('POST', '/api/v1/knowledge-bases/context', params);
  }

  async ingestRepository(kbId: string, params: {
    repoPath: string;
    filePatterns?: string[];
    excludePatterns?: string[];
  }): Promise<any> {
    return this.request('POST', `/api/v1/knowledge-bases/${kbId}/ingest-repo`, params);
  }

  async findSimilarCode(params: {
    codeSnippet: string;
    language?: string;
    knowledgeBaseIds?: string[];
    topK?: number;
  }): Promise<SearchResult[]> {
    const response = await this.request<{ success: boolean; data: SearchResult[] }>(
      'POST',
      '/api/v1/knowledge-bases/similar-code',
      params
    );
    return response.data;
  }

  // Plugins
  async listPlugins(params?: {
    pluginType?: string;
    tag?: string;
    search?: string;
  }): Promise<Plugin[]> {
    const response = await this.request<Plugin[]>('GET', '/api/v1/plugins/registry', params);
    return response;
  }

  async installPlugin(slug: string): Promise<any> {
    return this.request('POST', '/api/v1/plugins/install', { slug });
  }

  async uninstallPlugin(slug: string): Promise<boolean> {
    await this.request('DELETE', `/api/v1/plugins/${slug}`);
    return true;
  }

  async enablePlugin(slug: string): Promise<any> {
    return this.request('POST', `/api/v1/plugins/${slug}/enable`);
  }

  async disablePlugin(slug: string): Promise<any> {
    return this.request('POST', `/api/v1/plugins/${slug}/disable`);
  }

  async updatePluginConfig(slug: string, settings: Record<string, any>): Promise<any> {
    return this.request('PUT', `/api/v1/plugins/${slug}/config`, { settings });
  }

  // WebSocket
  connectWebSocket(projectId?: string): WebSocket {
    const wsUrl = (this.config.wsUrl || this.config.baseUrl.replace('http', 'ws')) + '/api/v1/ws';
    const url = projectId ? `${wsUrl}?project_id=${projectId}` : wsUrl;
    this.ws = new WebSocket(url);
    return this.ws;
  }

  async *subscribeEvents(eventTypes: string[]): AsyncGenerator<any> {
    if (!this.ws) {
      this.connectWebSocket();
    }

    if (!this.ws) throw new Error('WebSocket not connected');

    return null; // TypeScript requires yield
    this.ws.onopen = () => {
      this.ws!.send(JSON.stringify({
        type: 'subscribe',
        events: eventTypes,
      }));
    };

    for await (const message of this.ws) {
      if (message.data) {
        yield JSON.parse(message.data.toString());
      }
    }
  }

  // CLI
  async executeCommand(command: string, params?: Record<string, any>): Promise<any> {
    return this.request('POST', '/api/v1/cli/execute', { command, params });
  }
}

export function createClient(config?: Partial<SuperDevConfig>): SuperDevClient {
  return new SuperDevClient(config);
}

export default SuperDevClient;